Assets/Shopify/UIToolkit/Components/VariantOptionLookup.cs

```diff
@@ -0,0 +1,129 @@
+    ///     "node": {
+    ///         "title": "Blue / 1.5m",
+    ///         "selectedOptions": [
+    ///         {
+    ///             "name": "Color",
+    ///             "value": "Blue"
+    ///         },
+    ///         {
+    ///             "name": "Length",
+    ///             "value": "1.5m"
+    ///         }
+    ///         ]
+    ///     }
+    /// }
+    /// \endcode
+    /// 
+    /// Typically from a UI, we would want to select from 2 different dropdowns: 
+    /// a color picker and a size picker and not from a single Color/Size drop down.
+    ///
+    /// This class takes in a list of product variants and constructs an initial list of options
+    /// that the user can select from. Once a user selects an option, the developer can pass the
+    /// selected option to this class to get a list of the next set of options the user can select.
+    /// In this case, the initial set of options would be the colors (Blue/Green/Red). Once a user
+    /// selects their color, the developer can pass the color back into this class to get the list of
+    /// sizes to choose from (1m/2m/5m).
+    /// </summary>
+    public class VariantOptionLookup {
+        public List<SelectedOption> InitialOptions { get; private set; }
+        private List<ProductVariant> _variants;
+
+        /// <summary>
+        /// Takes in an initial list of product variants to build a list
+        /// of initial selection options to use. We assume that the first option type
+        /// listed in the array the GraphQL API returns is the initial option type.
+        /// </summary>
+        /// <param name="variants">Initial set of product variants with options.</param>
+        public VariantOptionLookup(List<ProductVariant> variants) {
+            InitialOptions = new List<SelectedOption>();
+
+            // No variants? No problem - just return;
+            if (variants.Count == 0) {
+                return;
+            }
+
+            // Is this a single variant/option product?
+            if (variants.Count == 1 && variants[0].title() == "Default Title") {
+                return;
+            }
+
+            _variants = variants;
+
+            AddInitialOptions(variants);
+        }
+
+        /// <summary>
+        /// Adds the given list of variants into our lookup table and adds the initial 
+        /// selection options to the initial list.
+        /// </summary>
+        /// <param name="moreVariants">An additional set of product variants.</param>
+        public void AddPageOfVariants(List<ProductVariant> moreVariants) {
+            _variants.AddRange(moreVariants);
+            AddInitialOptions(moreVariants);
+        }
+
+        /// <summary>
+        /// Returns the next set of selectable options given a selected option. For example,
+        /// if a user selects a Color option and the product supports selecting both a color 
+        /// and a size, this API will return a list of all available sizes to choose from.
```

Comment:
what if there are more than 2 selected options?

Reply:
I was wondering about this too. There seems to be some assumption that this is called each time a selector is used, and the list of further available options is updated in the subsequent options selection fields. That seems to break down though if the options are selected out of order.

Reply:
Just to close the loop on this, I'm going to rewrite this to take in a list of already-selected options and return a filtered list of all remaining available options. For example, when passing in nothing, it will return all of the available selections for each option. After selecting regardless of order it will produce the available selections for the remaining options. For example,